portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -136,10 +215,41 @@ private PreparedQuery getAllComments() {
     return datastore.prepare(query);
   }
 
+  /** 
+    * Returns all user comments that contain `search` in either
+    * the comment text or author name. 
+    * @param results All available comment entities in the database.
+    * @param comments An empty list to be filled. 
+    * @param search A string to filter the comments by. 
+    */
+  private ArrayList<Comment> getFilteredComments(PreparedQuery results, 
+    ArrayList<Comment> comments, String search) {
+    results.asIterable().forEach(commentEntity -> {
+      comments.add(Comment.fromEntity(commentEntity));
+    });
+
+    ArrayList<Comment> filteredComments = new ArrayList<Comment>();
+    Iterables.filter(comments, c -> satisfiesSearch(c, search))
+      .forEach(c -> filteredComments.add(c));
+    return filteredComments;
+  }
+
+  /** 
+    * Returns true if this comment contains the search string in the
+    * comment text or author name. 
+    */
+  private boolean satisfiesSearch(Comment comment, String search) {
+    if (Strings.isNullOrEmpty(search)) {
+      return true;
+    } else {
+      return comment.text.contains(search) || comment.name.contains(search);
+    }
+  }
+
   /** Returns JSON string representation of `data`. */
-  private String convertToJson(List<Comment> data) {
+  private String convertToJson(List<Comment> data, int pageToken) {
     Gson gson = new Gson(); 
     String json = gson.toJson(data);
-    return json;
+    return "[{\"pageToken\" : "+pageToken+" }, "+json+"]";
```

Comment:
Also, consider a generic datastructure so that the serialization doesn't have to be done manually.

Reply:
I like this idea --- would I be able to do this without making a new class/object to use, given that pageToken and data are different types? Would something like ArrayList<? extends object>.add(pageToken).add(data) work? I had assumed it would't since I'm not sure if ArrayList<Comment> extends object? I could potentially make a new object to hold these two variables for the purposes of converting to json without manually creating the string, if you don't mind the extra code! Let me know :)

Reply:
I'd encourage you to work with the team to figure it out :) I'm not sure on the exact answer, because we use a different framework internally.